proposals/csharp-9.0/nullable-reference-types-specification.md

```diff
@@ -285,62 +307,155 @@ If a `simple_name` is not classified as a value, its null state is "not null". O
 
+```c#
+var person = new Person();
+
+// The receiver is a tracked expression hence the member_access of the property 
+// is tracked as well 
+if (person.FirstName is not null)
+{
+    Use(person.FirstName);
+}
+
+// The return of an invocation is not a tracked expression hence the member_access
+// of the return is also not tracked
+if (GetAnonymous().FirstName is not null)
+{
+    // Warning: Cannot convert null literal to non-nullable reference type.
+    Use(GetAnonymous().FirstName);
+}
+
+void Use(string s) 
+{ 
+    // ...
+}
+
+public class Person
+{
+    public string? FirstName { get; set; }
+    public string? LastName { get; set; }
+
+    private static Person s_anonymous = new Person();
+    public static Person GetAnonymous() => s_anonymous;
+}
+```
+
 ### Invocation expressions
 
 If an `invocation_expression` invokes a member that is declared with one or more attributes for special null behavior, the null state is determined by those attributes. Otherwise the null state of the expression is the default null state of its type.
 
+The null state of an `invocation_expression` is not tracked by the compiler.
+
+```c#
+
+// The result of an invocation_expression is not tracked
+if (GetText() is not null)
+{
+    // Warning: Converting null literal or possible null value to non-nullable type.
+    string s = GetText();
+    // Warning: Dereference of a possibly null reference.
+    Use(s);
+}
+
+// Nullable friendly pattern
+if (GetText() is string s)
+{
+    Use(s);
+}
+
+string? GetText() => ... 
+Use(string s) {  }
+```
+
 ### Element access
 
 If an `element_access` invokes an indexer that is declared with one or more attributes for special null behavior, the null state is determined by those attributes. Otherwise the null state of the expression is the default null state of its type.
 
+```c#
+object?[] array = ...;
+if (array[0] != null)
+{
+    // Warning: Converting null literal or possible null value to non-nullable type.
+    object o = array[0];
+    // Warning: Dereference of a possibly null reference.
+    Console.WriteLine(o.ToString());
+}
+
+// Nullable friendly pattern
+if (array[0] is {} o)
+{
+    Console.WriteLine(o.ToString());
+}
+```
+
 ### Base access
 
 If `B` denotes the base type of the enclosing type, `base.I` has the same null state as `((B)this).I` and `base[E]` has the same null state as `((B)this)[E]`.
 
 ### Default expressions
 
-`default(T)` has the null state "not null" if `T` is known to be a nonnullable value type. Otherwise it has the null state "maybe default".
+`default(T)` has the null state based on the properties of the type `T`:
+
+- If the type is a *nonnullable* type then it has the null state "not null"
+- Else if the type is a type parameter then it has the null state "maybe default"
+- Else it has the null state "maybe null"
```

Comment:
- If the type is a *non-nullable value type* then it has the null state "not null"
- Else if the type is a type parameter that may accept either nullable reference type or non-nullable reference type arguments then it has the null state "maybe default"
- Else it has the null state "maybe null"

(it is feeling like we might want to have a shorthand for such type parameters which require this third state. Call them Heisenparameters if you want.)

Reply:
Why do you think me second bullet is incorrect here? Every case I could think of has "maybe default". The restrictions you added aren't necessary as far as I can see. Can you add a counter example to show why this isneeded?

Note: I've chosen *nonnullable* over *non-nullable* as that is the spelling the "Nullability of types" section uses.

Reply:
re: bullet 1: `string` is a nonnullable reference type, yet the flow state of `default(string)` is maybe-null. Maybe I overlooked the definition of *nonnullable type*?

re: bullet 2:
```cs
#nullable enable
void M<T>() where T : class
{
    T? t = default(T); // has maybe-null state
}
```

In the above scenario, we don't give a "maybe-default" state because there is no need to distinguish "maybe null, but is a valid value of type T" from "maybe null, and possibly an invalid value of type T" (maybe-default). Since we don't allow, for example, type parameter constraints that allows `string?` but not `string`, we could say that this state is required for any type parameter which accepts a nullable reference type argument.

Alternatively, we could collapse things down to say that "maybe default" states can be implicitly raised to "maybe null" states when the type doesn't require the usage of a "maybe default" state, but in this case, we might as well eliminate bullet 2 entirely and say in bullet 3: *else it has the null state "maybe default"*.